File: OshiLife/Services/MapService.swift
import Foundation
import UIKit

enum MapProvider: String, CaseIterable, Identifiable {
    case apple
    case google

    var id: Self { self }
}

enum MapServiceError: LocalizedError, Equatable {
    case emptyVenue
    case missingCoordinates
    case invalidMapURL

    var errorDescription: String? {
        switch self {
        case .emptyVenue: String(localized: "error.map_empty")
        case .missingCoordinates: String(localized: "error.map_coordinates")
        case .invalidMapURL: String(localized: "error.map_url")
        }
    }
}

@MainActor
struct MapService {
    func open(
        provider: MapProvider,
        venue: String,
        latitude: Double? = nil,
        longitude: Double? = nil
    ) throws {
        let trimmedVenue = venue.trimmingCharacters(in: .whitespacesAndNewlines)
        let url: URL
        if !trimmedVenue.isEmpty {
            url = switch provider {
            case .apple: try Self.appleMapsURL(venue: trimmedVenue)
            case .google: try Self.googleMapsURL(venue: trimmedVenue)
            }
        } else {
            let coordinateQuery = try Self.coordinateQuery(
                latitude: latitude,
                longitude: longitude
            )
            url = switch provider {
            case .apple: try Self.appleMapsURL(venue: coordinateQuery)
            case .google: try Self.googleMapsURL(venue: coordinateQuery)
            }
        }
        // Universal map links open an installed app and remain valid browser fallbacks.
        UIApplication.shared.open(url)
    }

    static func appleMapsURL(venue: String) throws -> URL {
        try mapsURL(
            scheme: "http",
            host: "maps.apple.com",
            path: "/",
            venue: venue
        )
    }

    static func googleMapsURL(venue: String) throws -> URL {
        try mapsURL(
            scheme: "https",
            host: "www.google.com",
            path: "/maps/search/",
            venue: venue,
            includesAPIParameter: true
        )
    }

    static func hasValidCoordinates(latitude: Double?, longitude: Double?) -> Bool {
        guard let latitude, let longitude else { return false }
        return (-90...90).contains(latitude) && (-180...180).contains(longitude)
    }

    static func googleMapsURL(latitude: Double?, longitude: Double?) throws -> URL {
        try googleMapsURL(venue: coordinateQuery(latitude: latitude, longitude: longitude))
    }

    private static func coordinateQuery(
        latitude: Double?,
        longitude: Double?
    ) throws -> String {
        guard hasValidCoordinates(latitude: latitude, longitude: longitude),
              let latitude,
              let longitude else {
            throw MapServiceError.missingCoordinates
        }
        return "\(latitude),\(longitude)"
    }

    private static func mapsURL(
        scheme: String,
        host: String,
        path: String,
        venue: String,
        includesAPIParameter: Bool = false
    ) throws -> URL {
        let venue = venue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !venue.isEmpty else {
            throw MapServiceError.emptyVenue
        }

        var components = URLComponents()
        components.scheme = scheme
        components.host = host
        components.path = path
        components.queryItems = includesAPIParameter
            ? [
                URLQueryItem(name: "api", value: "1"),
                URLQueryItem(name: "query", value: venue)
            ]
            : [URLQueryItem(name: "q", value: venue)]
        guard let url = components.url else { throw MapServiceError.invalidMapURL }
        return url
    }
}

File: OshiLifeAndroid/lib/core/utils/map_service.dart
/// Port of `OshiLife/Services/MapService.swift` for Android (plan §7.4):
/// name-first query, coordinates fallback. Apple Maps is replaced by a
/// generic `geo:` URI that opens the user's chosen maps app; the Google
/// Maps URL is identical to iOS. Callers launch the returned URIs with
/// url_launcher.
enum MapServiceErrorKind { emptyVenue, missingCoordinates }

class MapServiceException implements Exception {
  const MapServiceException(this.kind);

  final MapServiceErrorKind kind;

  /// iOS `errorDescription` localization keys.
  String get localizationKey => switch (kind) {
    MapServiceErrorKind.emptyVenue => 'error.map_empty',
    MapServiceErrorKind.missingCoordinates => 'error.map_coordinates',
  };
}

abstract final class MapService {
  static bool hasValidCoordinates(double? latitude, double? longitude) {
    if (latitude == null || longitude == null) return false;
    return latitude >= -90 &&
        latitude <= 90 &&
        longitude >= -180 &&
        longitude <= 180;
  }

  /// Same URL template as iOS `googleMapsURL(venue:)`.
  static Uri googleMapsUri({
    required String venue,
    double? latitude,
    double? longitude,
  }) {
    final query = _query(
      venue: venue,
      latitude: latitude,
      longitude: longitude,
    );
    return Uri.https('www.google.com', '/maps/search/', {
      'api': '1',
      'query': query,
    });
  }

  /// Android-generic map intent: opens the system chooser of installed
  /// maps apps.
  static Uri geoUri({
    required String venue,
    double? latitude,
    double? longitude,
  }) {
    final query = _query(
      venue: venue,
      latitude: latitude,
      longitude: longitude,
    );
    return Uri.parse('geo:0,0?q=${Uri.encodeComponent(query)}');
  }

  /// Name-first, coordinates fallback — the iOS `open(provider:...)` rule.
  static String _query({
    required String venue,
    double? latitude,
    double? longitude,
  }) {
    final trimmed = venue.trim();
    if (trimmed.isNotEmpty) return trimmed;
    if (!hasValidCoordinates(latitude, longitude)) {
      throw const MapServiceException(MapServiceErrorKind.missingCoordinates);
    }
    return '$latitude,$longitude';
  }
}
